pinot-controller/src/main/java/org/apache/pinot/controller/LeadControllerManager.java

```diff
@@ -0,0 +1,163 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.controller;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import org.apache.helix.model.ExternalView;
+import org.apache.pinot.common.config.TableNameBuilder;
+import org.apache.pinot.common.utils.helix.LeadControllerUtils;
+import org.apache.pinot.controller.helix.core.PinotHelixResourceManager;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static org.apache.pinot.common.utils.CommonConstants.Helix.LEAD_CONTROLLER_RESOURCE_NAME;
+import static org.apache.pinot.common.utils.CommonConstants.Helix.NUMBER_OF_PARTITIONS_IN_LEAD_CONTROLLER_RESOURCE;
+
+
+public class LeadControllerManager {
+  private static final Logger LOGGER = LoggerFactory.getLogger(LeadControllerManager.class);
+
+  private Map<Integer, Integer> _partitionIndexCache;
+  private PinotHelixResourceManager _pinotHelixResourceManager;
+  private volatile boolean _amIHelixLeader = false;
+
+  public LeadControllerManager() {
+    _partitionIndexCache = new ConcurrentHashMap<>();
+  }
+
+  public void registerResourceManager(PinotHelixResourceManager pinotHelixResourceManager) {
+    _pinotHelixResourceManager = pinotHelixResourceManager;
+  }
+
+  /**
+   * Check whether the current controller is the leader for the given table. Return true if current controller is the leader for this table.
+   * Otherwise check whether the current controller is helix leader.
+   * @param tableName table name with/without table type.
+   */
+  public boolean isLeaderForTable(String tableName) {
+    String rawTableName = TableNameBuilder.extractRawTableName(tableName);
+    int partitionIndex =
+        LeadControllerUtils.getPartitionIdForTable(rawTableName, NUMBER_OF_PARTITIONS_IN_LEAD_CONTROLLER_RESOURCE);
+    if (_partitionIndexCache.containsKey(partitionIndex)) {
```

Comment:
If there are 3 controllers, and controller A is leader for tables T1 and T2, and Controller B is leader for tables T3 and T4, and we call this method in contoller A for table T3 what happens?

Reply:
If that's the case, the current controller returns `false` since it isn't the leader for the given table, which is mentioned in the javadoc of this method.

Reply:
What if we are running in dual mode, and A happens to be helix leader?

Reply:
Refactored the code to use resource config to determine whether the resource is enabled.